snapshot/win/process_reader_win.cc

```diff
@@ -125,25 +126,102 @@ HANDLE OpenThread(
   return handle;
 }
 
+void DoStackWalk(ProcessReaderWin::Thread* thread,
+                 HANDLE process,
+                 HANDLE thread_handle,
+                 bool is_64_reading_32) {
+  if (is_64_reading_32) {
+    // TODO: we dont support it right away, maybe in the future
+    return;
+  }
+
+  STACKFRAME64 stack_frame;
+  memset(&stack_frame, 0, sizeof(stack_frame));
+
+  stack_frame.AddrPC.Mode = AddrModeFlat;
+  stack_frame.AddrFrame.Mode = AddrModeFlat;
+  stack_frame.AddrStack.Mode = AddrModeFlat;
+
+  int machine_type = IMAGE_FILE_MACHINE_I386;
+  LPVOID ctx = NULL;
+#if defined(ARCH_CPU_X86)
+  const CONTEXT* ctx_ = &thread->context.native;
+  stack_frame.AddrPC.Offset = ctx_->Eip;
+  stack_frame.AddrFrame.Offset = ctx_->Ebp;
+  stack_frame.AddrStack.Offset = ctx_->Esp;
+  ctx = (LPVOID)ctx_;
+#elif defined(ARCH_CPU_X86_64)
+  // if (!is_64_reading_32) {
+  machine_type = IMAGE_FILE_MACHINE_AMD64;
+
+  const CONTEXT* ctx_ = &thread->context.native;
+  stack_frame.AddrPC.Offset = ctx_->Rip;
+  stack_frame.AddrFrame.Offset = ctx_->Rbp;
+  stack_frame.AddrStack.Offset = ctx_->Rsp;
+  ctx = (LPVOID)ctx_;
+  // } else {
+  //   const WOW64_CONTEXT* ctx_ = &thread->context.wow64;
+  //   stack_frame.AddrPC.Offset = ctx_->Eip;
+  //   stack_frame.AddrFrame.Offset = ctx_->Ebp;
+  //   stack_frame.AddrStack.Offset = ctx_->Esp;
+  //   ctx = (LPVOID)ctx_;
+  // }
+
+// TODO: we dont support this right away, maybe in the future
+//#elif defined(ARCH_CPU_ARM64)
+//  machine_type = IMAGE_FILE_MACHINE_ARM64;
+#else
+#error Unsupported Windows Arch
+#endif  // ARCH_CPU_X86
+
+  char buffer[sizeof(SYMBOL_INFO) + MAX_SYM_NAME];
+  PSYMBOL_INFO pSymbol = (PSYMBOL_INFO)buffer;
```

Comment:
I don't understand what this does. Couldn't you just allocate a `PSYMBOL_INFO`?

Reply:
This is copied from here actually:
https://docs.microsoft.com/en-us/windows/win32/debug/retrieving-symbol-information-by-address
The only change I made is that I removed a multiply by `sizeof(TCHAR)` (which assumes wide-chars).

As the doc says:
> Because the name is variable in length, you must supply a buffer that is large enough to hold the name stored at the end of the SYMBOL_INFO structure.

The symbol bytes are written inline after the struct itself.

Reply:
Ok, I think I understand.